# M3 T02

## Exercici 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [4]:
import pandas as pd

df = pd.read_csv('/Users/jordi/Downloads/archive/DelayedFlights.csv' , index_col=0)
df = df.sample(n=200000, replace=False)
df

#Per començar importem la llibreria Pandas que és la que farem servir per treballar amb el dataset i els dataframes.
#El print el fem directament amb 'df' i no print('df') per fer-ho més visual.

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
6937389  2008     12          24          3   2144.0        2135     36.0   
2260388  2008      4          18          5   1656.0        1625   1831.0   
4698911  2008      8           5          2   1651.0        1515   1748.0   
1253427  2008      3          26          3   1929.0        1815   2255.0   
884523   2008      2          29          5   1836.0        1815   2115.0   
...       ...    ...         ...        ...      ...         ...      ...   
330718   2008      1          18          5   1241.0        1225   1458.0   
2357168  2008      4          29          2   1517.0        1505   1704.0   
4089842  2008      7          31          4   2004.0        1921   2254.0   
2227951  2008      4          25          5    938.0         916   1116.0   
5479250  2008     10          31          5   1309.0        1300   1649.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
6937389          30            AA       2020  ...    5.0     10.0          0   
2260388        1730            9E       5939  ...    8.0     29.0          0   
4698911        1630            9E       3779  ...    5.0     13.0          0   
1253427        2145            WN        567  ...    6.0     16.0          0   
884523         2037            US       1445  ...   23.0     35.0          0   
...             ...           ...        ...  ...    ...      ...        ...   
330718         1438            EV       4399  ...   18.0     10.0          0   
2357168        1648            CO        149  ...    8.0     19.0          0   
4089842        2155            9E       5769  ...    8.0     36.0          0   
2227951        1054            NW       1135  ...   11.0     14.0          0   
5479250        1642            XE       2105  ...    4.0     18.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
6937389                 N         0           NaN          NaN      NaN   
2260388                 N         0           0.0          0.0     30.0   
4698911                 N         0          78.0          0.0      0.0   
1253427                 N         0           0.0          0.0      0.0   
884523                  N         0          21.0          0.0     17.0   
...                   ...       ...           ...          ...      ...   
330718                  N         0          16.0          0.0      4.0   
2357168                 N         0           0.0          0.0      4.0   
4089842                 N         0           0.0          0.0     16.0   
2227951                 N         0          22.0          0.0      0.0   
5479250                 N         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
6937389            NaN                NaN  
2260388            0.0               31.0  
4698911            0.0                0.0  
1253427            0.0               70.0  
884523             0.0                0.0  
...                ...                ...  
330718             0.0                0.0  
2357168            0.0               12.0  
4089842            0.0               43.0  
2227951            0.0                0.0  
5479250            NaN                NaN  

[200000 rows x 29 columns]

# Anàlisi del dataset i les seves variables.

Sobre el dataset, aquest realitza un seguiment de la puntualitat dels vols nacionals als EUA operats per grans companyies aèries. La informació disponible analitza la quantitat de vols a temps, amb retard, cancel·lats i desviats de vols del 2008.

Es realitzarà una anàlisi del dataset proporcionat, el qual conté fins a 1.936.758 vols interns diferents als EUA per 2008 i les seves causes de retard, desviament i cancel·lació. Les dades provenen de la Oficina d'Estadístiques de Transport (BTS) del Departament de Transport dels EUA (DOT). Per treballar sobre ell, agafarem un sample de 200000 dades.

Les variables que hi podem trobar són les següents:

* **Year**: L'any, el 2008.

* **Month**: Mesos, de l'1 al 12.
* **DayofMonth**: Dia del mes, de l'1 al 31.
* **DayofWeek**: Dia de la setmana, de l'1 (dilluns) al 7 (diumenge).
* **DepTime**: hora de sortida real (local, hhmm).
* **CRSDepTime**: hora de sortida programada (local, hhmm).
* **ArrTime**: hora d'arribada real (local, hhmm).
* **CRSArrTime**: hora d'arribada programada (local, hhmm).
* **UniqueCarrier**: codi únic de la companyia aèria.
* **FlightNum**: número de vol.
* **TailNum**: número de cua de l'avió: matrícula de l'aeronau, identificador únic de l'aeronau.
* **ActualElapsedTime**: Temps real transcorregut en minuts.
* **CRSElapsedTime**: Temps transcorregut en minuts.
* **AirTime**: en minuts.
* **ArrDelay**: retard en l'arribada, en minuts: Un vol es compta com "a temps" si es va realitzar amb menys de 15 minuts de retard respecte a l'hora programada que es mostra en els Sistemes de Reserves Computaritzades (CRS en anglés) de les aerolínies.
* **DepDelay**: retard en la sortida, en minuts.
* **Origin**: codi d'origen de l'aeroport IATA.
* **Dest**: codi de destí d'aeroport IATA.
* **Distance**: Distància en milles.
* **TaxiIn**: Temps en minuts quan l'aeronau te autorització per entrar a pista.
* **TaxiOut**: Temps en minuts quan l'aeronau surt de pista.
* **Cancelled**: Si el vol va ser cancel·lat o no: 0 = no, 1 = sí.
* **CancellationCode**: motiu de la cancel·lació (A = transportista, B = clima, C = NAS, D = seguretat).
* **Diverted**: Desviats: 1 = sí, 0 = no.
* **CarrierDelay**: retard en minuts de l'aeronau per culpa de l'aerolínia. Per exemple: repostatge de l'aeronau, neteja de la mateixa, avaria, etc.
* **WeatherDelay**: retard en minuts a conseqüència de condicions climatològiques adverses que poden transcòrrer abans de la sortida de l'aeroport orígen, durant el vol, o a l'aeroport destí. Per exemple: tempestes elèctriques, grans nevades, etc.
* **NASDelay**: en minuts: el retard que està sota el control del Sistema Nacional de l'Espai Aèri (NAS) pot incloure: condicions climàtiques no extremes, operacions aeroportuàries, volum de tràfic pesat, control de tràfic aèri, etc.
* **SecurityDelay**: en minuts: el retard de seguretat es produeix per l'evacuació d'una terminal o vestíbul, tornar a embarcar una aeronau a causa d'una violació de la seguretat, equips de control inoperatius i/o llargues files de més de 29 minuts als controls de seguretat de l'aeroport.
* **LateAircraftDelay**: en minuts: Retard en l'arribada a un aeroport a causa de l'arribada tardana de la mateixa aeronau a un aeroport anterior. L'efecte dominó d'un retard anterior als aeroports posteriors es denomina propagació del retard.




In [5]:
df = df.drop(['Year','FlightNum','CancellationCode'], axis=1)
df

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
6937389     12          24          3   2144.0        2135     36.0   
2260388      4          18          5   1656.0        1625   1831.0   
4698911      8           5          2   1651.0        1515   1748.0   
1253427      3          26          3   1929.0        1815   2255.0   
884523       2          29          5   1836.0        1815   2115.0   
...        ...         ...        ...      ...         ...      ...   
330718       1          18          5   1241.0        1225   1458.0   
2357168      4          29          2   1517.0        1505   1704.0   
4089842      7          31          4   2004.0        1921   2254.0   
2227951      4          25          5    938.0         916   1116.0   
5479250     10          31          5   1309.0        1300   1649.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  Distance  \
6937389          30            AA  N469AA              112.0  ...       732   
2260388        1730            9E  89039E               95.0  ...       211   
4698911        1630            9E  87839E               57.0  ...       229   
1253427        2145            WN  N756SA              146.0  ...       997   
884523         2037            US  N763US              339.0  ...      2136   
...             ...           ...     ...                ...  ...       ...   
330718         1438            EV  N683BR              137.0  ...       640   
2357168        1648            CO  N30401              167.0  ...       965   
4089842        2155            9E  86599E              110.0  ...       441   
2227951        1054            NW  N676MC               98.0  ...       408   
5479250        1642            XE  N14125              160.0  ...      1042   

         TaxiIn  TaxiOut  Cancelled Diverted CarrierDelay  WeatherDelay  \
6937389     5.0     10.0          0        0          NaN           NaN   
2260388     8.0     29.0          0        0          0.0           0.0   
4698911     5.0     13.0          0        0         78.0           0.0   
1253427     6.0     16.0          0        0          0.0           0.0   
884523     23.0     35.0          0        0         21.0           0.0   
...         ...      ...        ...      ...          ...           ...   
330718     18.0     10.0          0        0         16.0           0.0   
2357168     8.0     19.0          0        0          0.0           0.0   
4089842     8.0     36.0          0        0          0.0           0.0   
2227951    11.0     14.0          0        0         22.0           0.0   
5479250     4.0     18.0          0        0          NaN           NaN   

         NASDelay  SecurityDelay  LateAircraftDelay  
6937389       NaN            NaN                NaN  
2260388      30.0            0.0               31.0  
4698911       0.0            0.0                0.0  
1253427       0.0            0.0               70.0  
884523       17.0            0.0                0.0  
...           ...            ...                ...  
330718        4.0            0.0                0.0  
2357168       4.0            0.0               12.0  
4089842      16.0            0.0               43.0  
2227951       0.0            0.0                0.0  
5479250       NaN            NaN                NaN  

[200000 rows x 26 columns]

He decidit eliminar les columnes "Year", donat que l'any és sempre el mateix i ja el sabem (2008); "FlightNum" perquè el número de vol ens es irrellevant per l'anàlisi que volem fer, ja que tot i que podriem pensar que podriem trobar patrons que es repeteixin en determinades rutes, una mateixa ruta pot (i sol) estar coberta per diferents avions, llavors hi hauria un petit biaix. Per últim eliminem "CancellationCode" ja que el codi de cancel·lació ens es indiferent per l'anàlisi que farem.
A continuació explico perquè he decidit deixar la resta, anant columna per columna:

* **Month**: Ens pot servir per analitzar si hi ha algun patró que influeixi que aquell mes hi hagin més retards, com podrien ser condicions climatològiques adverses perllongades, per exemple.

* **DayofMonth**: Podriem fixar-nos amb detall si hi un dia en concret va ocórrer alguna situació extrema, ja sigui climàtica, de seguretat, etc.

* **DayofWeek**: Aquí podriem analitzar si hi ha algun patró que es repeteixi un dia concret de la setmana.

* **DepTime**: Necessari per controlar l'hora de sortida real de l'aeronau.

* **CRSDepTime**: Necessitem saber l'hora programada per comparar-la amb la real i poder calcul·lar la diferència real en minuts.

* El mateix passa amb les columnes d'arribada **ArrTime** i **CRSArrTime**.

* **UniqueCarrier**: Ens es útil per comprovar si hi ha aerolínies que acumulin més retards que d'altres, podent extreure la conclusió que la gestió te a veure amb els retards.

* **TailNum**: Podriem fer un cribatge i comprovar si una mateixa aeronau acumula més retards que d'altres, depenent del model, antiguitat, possibles avaries més sovint, etc.

* **ActualElapsedTime**: Necessari per saber el temps real transcorregut.

* **Distance**: Ens serveix per veure si hi ha influència en el radi de les rutes. Com per exemple podriem pensar que a major distància, més opcions de trobar climatologia variada i/o adversa.

* Les columnes **TaxiIn** i **TaxiOut** per control·lar el temps real des de que l'aeronau te autorització a enlairar-se pero està bloquejada a pista (potser per cua a pista, climatologia, raons de seguretat…) a quan s'enlaira realment.

* Les columnes **Cancelled** i **Diverted** per comprovar si hi ha hagut cancelació o desviació del vol.

* **CarrierDelay**: Per veure el retard causat per l'aerolínia.

* **WeatherDelay**: Temps de retard causat per condicions climatològiques.

* **NASDelay**: Per comprovar el temps de retard causat per tancament d'espai aeri.

* **SecurityDelay**: Retard causat per problemes de seguretat.

* **LateAircraftDelay**: Ens és útil per saber si el retard ve causat perquè l'aeronau ja s'havia enlairat tard degut a un aterrament amb retard a l'aeroport orígen.

In [6]:
df.info()
#Exploració i tipus de dades que conté el dataframe.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 6937389 to 5479250
Data columns (total 26 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Month              200000 non-null  int64  
 1   DayofMonth         200000 non-null  int64  
 2   DayOfWeek          200000 non-null  int64  
 3   DepTime            200000 non-null  float64
 4   CRSDepTime         200000 non-null  int64  
 5   ArrTime            199264 non-null  float64
 6   CRSArrTime         200000 non-null  int64  
 7   UniqueCarrier      200000 non-null  object 
 8   TailNum            199999 non-null  object 
 9   ActualElapsedTime  199139 non-null  float64
 10  CRSElapsedTime     199981 non-null  float64
 11  AirTime            199139 non-null  float64
 12  ArrDelay           199139 non-null  float64
 13  DepDelay           200000 non-null  float64
 14  Origin             200000 non-null  object 
 15  Dest               200000 non-null  object 


In [7]:
df.head(20)
#Cridem els 20 primers valors.

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
6937389     12          24          3   2144.0        2135     36.0   
2260388      4          18          5   1656.0        1625   1831.0   
4698911      8           5          2   1651.0        1515   1748.0   
1253427      3          26          3   1929.0        1815   2255.0   
884523       2          29          5   1836.0        1815   2115.0   
3496008      6          11          3   1127.0        1120   1440.0   
2512023      5          15          4   1110.0        1103   1247.0   
1138797      2           7          4   1116.0        1045   1448.0   
6090523     11           7          5   1952.0        1915   2153.0   
4119261      7           3          4   1622.0        1550   1806.0   
3464043      6           8          7   1841.0        1750   2041.0   
1679660      3          24          1   1720.0        1640   1943.0   
4816735      8          24          7   1728.0        1722   1941.0   
552342       1          11          5   1439.0        1430   1551.0   
769769       2           4          1   1301.0        1235   1451.0   
3678604      7          25          5   1812.0        1745   1859.0   
6861789     12          15          1   1715.0        1650   1830.0   
1276854      3           5          3   1702.0        1445   2023.0   
6270678     11          26          3   1648.0        1631   1944.0   
3045923      6          17          2   1939.0        1930   2054.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  Distance  \
6937389          30            AA  N469AA              112.0  ...       732   
2260388        1730            9E  89039E               95.0  ...       211   
4698911        1630            9E  87839E               57.0  ...       229   
1253427        2145            WN  N756SA              146.0  ...       997   
884523         2037            US  N763US              339.0  ...      2136   
3496008        1420            AA  N5EMAA              193.0  ...      1097   
2512023        1235            XE  N15527              157.0  ...       871   
1138797        1443            CO  N61304              152.0  ...      1235   
6090523        2115            OO  N498CA               61.0  ...       238   
4119261        1745            AA  N498AA              164.0  ...      1005   
3464043        1952            9E  80249E              120.0  ...       616   
1679660        1910            AA  N469AA              203.0  ...      1192   
4816735        1922            CO  N73283              133.0  ...       416   
552342         1550            B6  N561JB               72.0  ...       353   
769769         1422            OO  N866AS              110.0  ...       601   
3678604        1835            WN  N614SW               47.0  ...       187   
6861789        1757            NW  N316US              135.0  ...       680   
1276854        1813            XE  N26549              141.0  ...       956   
6270678        1936            FL  N285AT              116.0  ...       732   
3045923        2045            WN  N333SW               75.0  ...       389   

         TaxiIn  TaxiOut  Cancelled Diverted CarrierDelay  WeatherDelay  \
6937389     5.0     10.0          0        0          NaN           NaN   
2260388     8.0     29.0          0        0          0.0           0.0   
4698911     5.0     13.0          0        0         78.0           0.0   
1253427     6.0     16.0          0        0          0.0           0.0   
884523     23.0     35.0          0        0         21.0           0.0   
3496008     7.0     38.0          0        0          2.0           0.0   
2512023     4.0     11.0          0        0          NaN           NaN   
1138797     5.0     13.0          0        0          NaN           NaN   
6090523     7.0      9.0          0        0          6.0           0.0   
4119261    10.0     10.0          0        0          0.0           0.0   
3464043     6.0     14.0        

# Exercici 2
Fes un informe complet del dataset:

* Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

* Troba quantes dades faltants hi ha per columna.
* Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
* Fes una taula de les aerolínies amb més endarreriments acumulats.
* Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

Aporta allò que consideris rellevant.


In [8]:

aerolinia = df["UniqueCarrier"]
#fem servir el mètode value_counts per veure les aerolínies que més i menys es repeteixen
aerolinia_suma = aerolinia.value_counts()
#seleccionem la primera aerolínia del total per veure la que acumula més retards
mes_retards = aerolinia_suma.index[0]
#seleccionem l'última aerolínia del total per veure la que acumula menys retards
menys_retards = aerolinia_suma.index[-1]

print (aerolinia_suma)
print ("L'aerolínia amb més retards és:", mes_retards)
print("L'aerolínia amb menys retards és:", menys_retards)


WN    39213
AA    19742
UA    14737
MQ    14606
OO    13546
DL    11870
XE    10672
US    10190
CO    10144
EV     8430
NW     8195
FL     7420
YV     6920
B6     5755
OH     5448
9E     5287
AS     4095
F9     2889
HA      765
AQ       76
Name: UniqueCarrier, dtype: int64
L'aerolínia amb més retards és: WN
L'aerolínia amb menys retards és: AQ


D'aquest petit anàlisi, si comprovem el codi IATA de les aerolínies, trobem que la que presenta més retards és Southwest Airlines (WN) una aerolínia de baix cost amb gran presència al llarg de tot el territori dels EUA, amb base a diferents ciutats on les condicions climatològiques solen ser adverses: Chicago molt de vent; Fort Lauderdale a Florida on pateixen temporades d'huracans, etc.
Per una altra banda, la que menys, Aloha Airlines (AQ) és una companyia que opera pocs vols, el qual ja fa incrementar la possibilitat de tenir menys retards pel pur volum de passatgers i disponibilitat d'aeronaus, així com s'afavoreix del bon clima d'on te la base operacional, situada a Hawaii, on gaudeixen de bon temps la gran majoria de l'any.

In [21]:
#faig una variable amb les columnes "AirTime" i "Distance"
df = df.drop(df[df["AirTime"]==0].index)
columnes_interes = ["AirTime", "Distance","DepDelay", "ArrDelay"]
#demanem un resum estadístic amb el mètode .describe()
df[columnes_interes].describe()

AirTime       Distance       DepDelay       ArrDelay
count  199136.000000  199997.000000  199997.000000  199136.000000
mean      108.519685     767.702791      43.139507      42.155562
std        68.940575     576.612146      53.091019      56.453467
min         1.000000      11.000000       6.000000     -63.000000
25%        58.000000     338.000000      12.000000       9.000000
50%        90.000000     607.000000      24.000000      24.000000
75%       137.000000    1005.000000      53.000000      56.000000
max       664.000000    4962.000000    1403.000000    1395.000000

Aquí he eliminat les files amb valor 0 de la columna "AirTime", per no esbiaixar la mitjana. Llavors podem treure de conclusió que:
* La mitjana de temps de vol esta en 108,52 minuts i la distància de vol en 767,70 milles.
* El 50% dels vols són de 607 milles de distància.
* El temps mijtà de retard son 42,16 minuts a l'arribada, sortint uns 43,13 minuts tard de l'aeroport orígen.
També podem apreciar com:
* La meitat dels vols tenen una duració de 90 minuts, és a dir, curt radi, i un 75% no més de 137 minuts. 
Podem apreciar possibles errors a les dades també, ja que mostra que el mínim es d'1 minut, o 11 milles de vol, el qual no acostuma a passar en vols comercials, tret d'alguns de molt petit radi entre illes per moure's per arxipèlags on les vies de comunicació son més complicades i escases. 
Donat que als EUA aquesta situació es dona a Alaska, a continuació ho comprovarem detectant el vol i mirant l'aeroport orígen i destí per discernir si es tracta d'un error a les dades o si és possible. 
* La desviació típica ens indica que de les dades obtingudes un 68,94% estan vora el voltant de la mitjana i no existeix una majoria dispersa.


In [16]:
# Seleccionem la columna d'interès
columna = df["AirTime"]

# Obtenim l'índex de la columna amb valor mínim
min_index = columna.idxmin()

# Seleccionem la fila al dataframe original i mostrem les dades del vol a pantalla.
min_fila = df.loc[min_index]
min_fila

Month                     4
DayofMonth               25
DayOfWeek                 5
DepTime              2110.0
CRSDepTime             1930
ArrTime              2139.0
CRSArrTime             2007
UniqueCarrier            OO
TailNum              N945SW
ActualElapsedTime      29.0
CRSElapsedTime         37.0
AirTime                 1.0
ArrDelay               92.0
DepDelay              100.0
Origin                  DEN
Dest                    COS
Distance                 72
TaxiIn                  4.0
TaxiOut                24.0
Cancelled                 0
Diverted                  0
CarrierDelay            0.0
WeatherDelay            0.0
NASDelay                0.0
SecurityDelay           0.0
LateAircraftDelay      92.0
Name: 1975668, dtype: object

Gràcies a que hem aïllat les dades del vol en concret, podem veure com efectivament es tracta d'un error. Només comprovant l'hora de sortida i d'aterratge ja veiem que no encaixa amb el temps que l'avió hauria d'haver estat enlairat, així com per la distància que trobem entre l'aeroport de Denver i Colorado Springs, malgrat que la distància és de només 72 milles.
Llavors, tornant a l'anàlisi fet anteriorment, podem confirmar que hi ha certs errors al dataframe que poden esbiaixar els resultats.

In [17]:
# Seleccionem la columna d'interès
columna2 = df["Distance"]

# Obtenim l'índex de la columna amb valor mínim
min_index2 = columna2.idxmin()

# Seleccionem la fila al dataframe original i mostrem les dades del vol a pantalla.
min_fila2 = df.loc[min_index2]
min_fila2

Month                     5
DayofMonth               15
DayOfWeek                 4
DepTime              2037.0
CRSDepTime             1800
ArrTime              2125.0
CRSArrTime             1900
UniqueCarrier            OH
TailNum              N806CA
ActualElapsedTime      48.0
CRSElapsedTime         60.0
AirTime                31.0
ArrDelay              145.0
DepDelay              157.0
Origin                  JFK
Dest                    LGA
Distance                 11
TaxiIn                 10.0
TaxiOut                 7.0
Cancelled                 0
Diverted                  0
CarrierDelay          145.0
WeatherDelay            0.0
NASDelay                0.0
SecurityDelay           0.0
LateAircraftDelay       0.0
Name: 2547298, dtype: object

Repetim procés per comprovar si la distància mínima és correcta i en aquest cas confirmem que sí és correcta, ja que ambdós estan situats a l'estat i ciutat de Nova York. 

Nogensmenys, donat la distància de només 11 milles, i que ens indiquen que l'avió va estar volant 31 minuts, ens fa pensar que es tracti d'algun vol privat o de llançadera operat per una companyia pública. Podriem valorar també l'opció de que deixès els passatgers al JFK (origen) i dormís a l'hangar a La Guardia (destí) Comprovem el codi IATA de l'aerolínia, OH, i veiem que pertany a PSA Airlines, companyia que opera amb jets privats amb vols domèstics. Malauradament (però el planeta ho agraeix), comprovem que aquesta ruta avui dia no existeix, el que ens deixa en una posició d'incertesa sobre si era una ruta real, una reubicació per aparcar l'aeronau, o un simple enllaç.

In [18]:
df.describe()
#Anàlisi estadístic bàsic de totes les variables.

Month     DayofMonth      DayOfWeek        DepTime  \
count  199997.000000  199997.000000  199997.000000  199997.000000   
mean        6.112037      15.728786       3.982010    1518.625059   
std         3.479055       8.782720       1.995063     449.741760   
min         1.000000       1.000000       1.000000       1.000000   
25%         3.000000       8.000000       2.000000    1203.000000   
50%         6.000000      16.000000       4.000000    1545.000000   
75%         9.000000      23.000000       6.000000    1900.000000   
max        12.000000      31.000000       7.000000    2400.000000   

          CRSDepTime        ArrTime     CRSArrTime  ActualElapsedTime  \
count  199997.000000  199261.000000  199997.000000      199136.000000   
mean     1467.381431    1611.083845    1634.912874         133.532922   
std       424.152978     547.018359     463.310870          72.287125   
min         1.000000       1.000000       0.000000          17.000000   
25%      1135.000000    1317.000000    1325.000000          80.000000   
50%      1510.000000    1716.000000    1706.000000         116.000000   
75%      1815.000000    2030.000000    2014.000000         165.000000   
max      2359.000000    2400.000000    2400.000000         711.000000   

       CRSElapsedTime        AirTime  ...       Distance         TaxiIn  \
count   199978.000000  199136.000000  ...  199997.000000  199261.000000   
mean       134.534514     108.519685  ...     767.702791       6.800714   
std         71.664993      68.940575  ...     576.612146       5.207220   
min          1.000000       1.000000  ...      11.000000       0.000000   
25%         82.000000      58.000000  ...     338.000000       4.000000   
50%        116.000000      90.000000  ...     607.000000       6.000000   
75%        165.000000     137.000000  ...    1005.000000       8.000000   
max        660.000000     664.000000  ...    4962.000000     176.000000   

             TaxiOut      Cancelled       Diverted   CarrierDelay  \
count  199952.000000  199997.000000  199997.000000  128827.000000   
mean       18.228160       0.000325       0.003980      19.158888   
std        14.280886       0.018025       0.062962      42.833569   
min         0.000000       0.000000       0.000000       0.000000   
25%        10.000000       0.000000       0.000000       0.000000   
50%        14.000000       0.000000       0.000000       2.000000   
75%        21.000000       0.000000       0.000000      21.000000   
max       379.000000       1.000000       1.000000    1344.000000   

        WeatherDelay       NASDelay  SecurityDelay  LateAircraftDelay  
count  128827.000000  128827.000000  128827.000000      128827.000000  
mean        3.780395      15.028038       0.089461          25.160518  
std        22.591036      33.545686       1.906888          41.885962  
min         0.000000       0.000000       0.000000           0.000000  
25%         0.000000       0.000000       0.000000           0.000000  
50%         0.000000       2.000000       0.000000           8.000000  
75%         0.000000      15.000000       0.000000          33.000000  
max      1297.000000     890.000000     254.000000        1143.000000  

[8 rows x 22 columns]

In [23]:
dades_nules = df.isna().sum()
dades_nules

#localitzem les dades nul·les

Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                736
CRSArrTime               0
UniqueCarrier            0
TailNum                  1
ActualElapsedTime      861
CRSElapsedTime          19
AirTime                861
ArrDelay               861
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 736
TaxiOut                 45
Cancelled                0
Diverted                 0
CarrierDelay         71170
WeatherDelay         71170
NASDelay             71170
SecurityDelay        71170
LateAircraftDelay    71170
dtype: int64

In [26]:
df.dropna(inplace=True)
df.shape
df
#netegem el dataset eliminant dades nul·les i NaNs.

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
2260388      4          18          5   1656.0        1625   1831.0   
4698911      8           5          2   1651.0        1515   1748.0   
1253427      3          26          3   1929.0        1815   2255.0   
884523       2          29          5   1836.0        1815   2115.0   
3496008      6          11          3   1127.0        1120   1440.0   
...        ...         ...        ...      ...         ...      ...   
1035055      2          24          7   1005.0         845   1119.0   
330718       1          18          5   1241.0        1225   1458.0   
2357168      4          29          2   1517.0        1505   1704.0   
4089842      7          31          4   2004.0        1921   2254.0   
2227951      4          25          5    938.0         916   1116.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  Distance  \
2260388        1730            9E  89039E               95.0  ...       211   
4698911        1630            9E  87839E               57.0  ...       229   
1253427        2145            WN  N756SA              146.0  ...       997   
884523         2037            US  N763US              339.0  ...      2136   
3496008        1420            AA  N5EMAA              193.0  ...      1097   
...             ...           ...     ...                ...  ...       ...   
1035055         940            9E  89039E              134.0  ...       554   
330718         1438            EV  N683BR              137.0  ...       640   
2357168        1648            CO  N30401              167.0  ...       965   
4089842        2155            9E  86599E              110.0  ...       441   
2227951        1054            NW  N676MC               98.0  ...       408   

         TaxiIn  TaxiOut  Cancelled Diverted CarrierDelay  WeatherDelay  \
2260388     8.0     29.0          0        0          0.0           0.0   
4698911     5.0     13.0          0        0         78.0           0.0   
1253427     6.0     16.0          0        0          0.0           0.0   
884523     23.0     35.0          0        0         21.0           0.0   
3496008     7.0     38.0          0        0          2.0           0.0   
...         ...      ...        ...      ...          ...           ...   
1035055    10.0     49.0          0        0         80.0           0.0   
330718     18.0     10.0          0        0         16.0           0.0   
2357168     8.0     19.0          0        0          0.0           0.0   
4089842     8.0     36.0          0        0          0.0           0.0   
2227951    11.0     14.0          0        0         22.0           0.0   

         NASDelay  SecurityDelay  LateAircraftDelay  
2260388      30.0            0.0               31.0  
4698911       0.0            0.0                0.0  
1253427       0.0            0.0               70.0  
884523       17.0            0.0                0.0  
3496008      13.0            0.0                5.0  
...           ...            ...                ...  
1035055      19.0            0.0                0.0  
330718        4.0            0.0                0.0  
2357168       4.0            0.0               12.0  
4089842      16.0            0.0               43.0  
2227951       0.0            0.0                0.0  

[128827 rows x 26 columns]

In [38]:
import numpy as np
df['Distance_km'] = df['Distance'] * 1.60934 
#Convertim milles a kilòmetres
df['AirTime'].replace(0, np.nan, inplace = True)
df['AirTime_Hores'] = df['AirTime'] / 60 
#Convertim minuts a hores.
df['Velocitat_Mitjana'] = df['Distance_km'] / df['AirTime_Hores']
#Afegim columna velocitat mitjana.
df[['Distance', 'AirTime_Hores', 'Velocitat_Mitjana']].head(20)
#Mostrem els primers 20 valors.

Distance  AirTime_Hores  Velocitat_Mitjana
2260388       211       0.966667         351.280076
4698911       229       0.650000         566.982862
1253427       997       2.066667         776.376765
884523       2136       4.683333         733.996493
3496008      1097       2.466667         715.721343
6090523       238       0.750000         510.697227
4119261      1005       2.400000         673.911125
3464043       616       1.666667         594.812064
1679660      1192       2.683333         714.906812
4816735       416       1.300000         514.988800
769769        601       1.466667         659.463641
3678604       187       0.566667         531.082200
6861789       680       1.866667         586.259571
1276854       956       2.050000         750.501971
2422401       197       0.550000         576.436327
4466376       783       1.933333         651.782700
843806        629       1.500000         674.849907
1579732       364       1.166667         502.114080
15198         180       0.550000         526.693091
4492955       414       1.200000         555.222300

In [34]:
df['Aterratge_Retard'] = df['ArrDelay'] > 0
df[['UniqueCarrier','Aterratge_Retard']].head(20)
#Afegim amb un True si el vol aterra tard. Mostrem els primers 20 valors.

UniqueCarrier  Aterratge_Retard
2260388            9E              True
4698911            9E              True
1253427            WN              True
884523             US              True
3496008            AA              True
6090523            OO              True
4119261            AA              True
3464043            9E              True
1679660            AA              True
4816735            CO              True
769769             OO              True
3678604            WN              True
6861789            NW              True
1276854            XE              True
2422401            WN              True
4466376            UA              True
843806             UA              True
1579732            MQ              True
15198              WN              True
4492955            UA              True

In [41]:
df['Enlairament_Retard'] = df['DepDelay'] > 0
df[['Distance', 'AirTime_Hores', 'Velocitat_Mitjana', 'Aterratge_Retard', 'Enlairament_Retard']].head(20)
#Creem un altre booleà per comprovar si l'aeronau s'ha enlairat amb retard. Mostrem els 20 primers valors de les variables anteriors + la nova.

Distance  AirTime_Hores  Velocitat_Mitjana  Aterratge_Retard  \
2260388       211       0.966667         351.280076              True   
4698911       229       0.650000         566.982862              True   
1253427       997       2.066667         776.376765              True   
884523       2136       4.683333         733.996493              True   
3496008      1097       2.466667         715.721343              True   
6090523       238       0.750000         510.697227              True   
4119261      1005       2.400000         673.911125              True   
3464043       616       1.666667         594.812064              True   
1679660      1192       2.683333         714.906812              True   
4816735       416       1.300000         514.988800              True   
769769        601       1.466667         659.463641              True   
3678604       187       0.566667         531.082200              True   
6861789       680       1.866667         586.259571              True   
1276854       956       2.050000         750.501971              True   
2422401       197       0.550000         576.436327              True   
4466376       783       1.933333         651.782700              True   
843806        629       1.500000         674.849907              True   
1579732       364       1.166667         502.114080              True   
15198         180       0.550000         526.693091              True   
4492955       414       1.200000         555.222300              True   

         Enlairament_Retard  
2260388                True  
4698911                True  
1253427                True  
884523                 True  
3496008                True  
6090523                True  
4119261                True  
3464043                True  
1679660                True  
4816735                True  
769769                 True  
3678604                True  
6861789                True  
1276854                True  
2422401                True  
4466376                True  
843806                 True  
1579732                True  
15198                  True  
4492955                True

In [42]:
#Rescato la taula de les aerolínies amb més i menys retards presentada anteriorment.

aerolinia = df["UniqueCarrier"]
#fem servir el mètode value_counts per veure les aerolínies que més i menys es repeteixen
aerolinia_suma = aerolinia.value_counts()
#seleccionem la primera aerolínia del total per veure la que acumula més retards
mes_retards = aerolinia_suma.index[0]
#seleccionem l'última aerolínia del total per veure la que acumula menys retards
menys_retards = aerolinia_suma.index[-1]

print (aerolinia_suma)
print ("L'aerolínia amb més retards és:", mes_retards)
print("L'aerolínia amb menys retards és:", menys_retards)

WN    21087
AA    13622
MQ    10021
UA     9880
OO     9118
DL     7435
XE     7428
US     6088
CO     6027
EV     5879
NW     5700
YV     5257
FL     4933
OH     4060
B6     3958
9E     3661
AS     2529
F9     1666
HA      451
AQ       27
Name: UniqueCarrier, dtype: int64
L'aerolínia amb més retards és: WN
L'aerolínia amb menys retards és: AQ


In [44]:
vols_llargs = df.sort_values(by= ['Distance'], ascending = False)
vols_llargs[:20] 
#Presentem els 20 vols més llargs.

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
6981772     12          17          3   1443.0        1335   2036.0   
577639       1           9          3   1448.0        1320   2055.0   
6426792     11          30          7   1455.0        1335   2037.0   
1139464      2          21          4   1337.0        1320   2014.0   
2351025      4          17          4   1508.0        1335   1938.0   
1742567      3          21          5   1345.0        1320   1936.0   
570350       1          22          2   1441.0        1320   2016.0   
4199498      7          30          3   1341.0        1335   1831.0   
1145965      2           6          3   2301.0        2245   1325.0   
6985264     12          12          5   1418.0        1335   1951.0   
5910090     10          19          7   1347.0        1335   1825.0   
6986029     12          11          4   1503.0        1335   2028.0   
2966111      5          15          4   1631.0        1335   2152.0   
4814977      8          26          2   2016.0        1955   1216.0   
4198600      7          31          4   1356.0        1335   1827.0   
4204661      7          12          6   1426.0        1335   1844.0   
1742533      3          18          2   2207.0        2145   1326.0   
2353671      4          25          5   1345.0        1335   1826.0   
2969130      5           9          5   1414.0        1335   1843.0   
7007224     12          11          4   1720.0        1650    708.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  \
6981772        1935            CO  N66051              653.0  ...   
577639         1920            CO  N66051              667.0  ...   
6426792        1935            CO  N69059              642.0  ...   
1139464        1920            CO  N67058              697.0  ...   
2351025        1810            CO  N76054              630.0  ...   
1742567        1820            CO  N66051              711.0  ...   
570350         1920            CO  N78060              635.0  ...   
4199498        1810            CO  N66057              650.0  ...   
1145965        1310            CO  N67058              564.0  ...   
6985264        1935            CO  N76055              633.0  ...   
5910090        1810            CO  N76062              638.0  ...   
6986029        1935            CO  N59053              625.0  ...   
2966111        1810            CO  N76054              681.0  ...   
4814977        1130            CO  N66056              600.0  ...   
4198600        1810            CO  N66056              631.0  ...   
4204661        1810            CO  N67052              618.0  ...   
1742533        1310            CO  N66056              559.0  ...   
2353671        1810            CO  N69059              641.0  ...   
2969130        1810            CO  N59053              629.0  ...   
7007224         632            DL  N829MH              528.0  ...   

         SecurityDelay  LateAircraftDelay  AirTime_H  Average_Speed  \
6981772            0.0               61.0  10.350000     771.550249   
577639             0.0               85.0  10.466667     762.950167   
6426792            0.0               60.0  10.266667     777.812832   
1139464            0.0                0.0  11.066667     721.585399   
2351025            0.0                0.0  10.016667     797.225798   
1742567            0.0                0.0  10.916667     731.500313   
570350             0.0               56.0  10.116667     789.345477   
4199498            0.0                0.0  10.166667     785.463450   
1145965            0.0                0.0   9.000000     887.282787   
6985264            0.0               16.0  10.100000     790.648028   
5910090            0.0                0.0  10.083333     791.954884   
6986029            0.0                0.0   9.983333     799.887654   
2966111            0.0              173.0  10.750000     742.841403   
4814977            0.0                0.0   9.300000     858.660761   
41

In [47]:
vols_mes_retard = df.sort_values(by = ['ArrDelay'], ascending = False) 
vols_mes_retard[:20] 
#Mostrem els 20 vols amb més retard ordenant inversament.

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
6868167     12          17          3   1048.0        1125   1748.0   
3435921      6          16          1   1258.0        1420   1753.0   
6916589     12          17          3    908.0        1355   1627.0   
2820874      5          11          7    637.0         939    925.0   
5226263      9          21          7   1300.0        1633   1451.0   
1612481      3          28          5    713.0        1048    939.0   
1022387      2          19          2    652.0        1110    942.0   
1095815      2           7          4    805.0        1150   1116.0   
1018915      2          17          7   1253.0        1745   1551.0   
4092707      7           2          3   1412.0        1925   1512.0   
2823103      5           3          6    910.0        1412   1420.0   
1071476      2           5          2    900.0        1510   1038.0   
4689050      8          14          4    701.0        1240    828.0   
4494069      8           4          1    754.0        1444   1209.0   
1026196      2          17          7    741.0        1442    921.0   
6862436     12          20          6   1602.0        2320   2356.0   
5313007      9           7          7    751.0        1450   1247.0   
1010406      2          22          5    959.0        1800   1240.0   
1031259      2           6          3   1100.0        1909   1130.0   
2261771      4           4          5   1312.0        2115   2120.0   

         CRSArrTime UniqueCarrier TailNum  ActualElapsedTime  ...  \
6868167        1833            NW  N557NW              240.0  ...   
3435921        1929            NW  N501US              175.0  ...   
6916589        1835            AA  N610AA              319.0  ...   
2820874        1222            NW  N521US              168.0  ...   
5226263        1812            NW  N532US              111.0  ...   
1612481        1315            NW  N520US              206.0  ...   
1022387        1339            NW  N987US              170.0  ...   
1095815        1520            AA  N236AA              131.0  ...   
1018915        2022            NW   N9338              178.0  ...   
4092707        2009            9E  89219E               60.0  ...   
2823103        1927            NW  N355NW              190.0  ...   
1071476        1630            AA  N520AA               98.0  ...   
4689050        1436            NW  N783NC               87.0  ...   
4494069        1859            UA  N406UA              315.0  ...   
1026196        1615            NW  N360NB              160.0  ...   
6862436         650            NW  N586NW              354.0  ...   
5313007        1955            AA  N5CAAA              176.0  ...   
1010406        2025            NW  N401EA              161.0  ...   
1031259        1921            NW  N778NC               90.0  ...   
2261771         515            AA  N343AA              308.0  ...   

         SecurityDelay  LateAircraftDelay  AirTime_H  Average_Speed  \
6868167            0.0               98.0   3.066667     918.373370   
3435921            0.0                0.0   2.650000     789.487547   
6916589            0.0                0.0   2.066667     821.542113   
2820874            0.0                0.0   2.483333     637.039417   
5226263            0.0               59.0   1.050000     625.343543   
1612481            0.0                0.0   3.000000     701.135793   
1022387            0.0              106.0   2.050000     657.866790   
1095815            0.0              135.0   1.866667     851.800671   
1018915            0.0             1143.0   1.983333     671.865640   
4092707            0.0                0.0   0.383333     235.103583   
2823103            0.0                7.0   2.866667     800.553084   
1071476            0.0                0.0   0.933333     565.568057   
4689050            0.0                0.0   1.250000     643.736000   
4494069            0.0              265.0   4.850000     687.536594   
10

In [51]:
vols_dia = df['DayOfWeek'].value_counts().sort_values(ascending=False)
vols_dia
#Mostrem els vols més repetits per dia. Sent de l'1 al 7 dilluns - diumenge a l'índex.

5    22116
1    19318
4    19292
7    18863
2    17600
3    17401
6    14237
Name: DayOfWeek, dtype: int64

# Exercici 3

Exporta el dataset net i amb les noves columnes a Excel.

In [52]:
df.to_excel('dataset.xlsx', index = False)

ModuleNotFoundError: No module named 'openpyxl'